### PR TITLE
Bug 1679380 - Implement the counter metric type

### PR DESCRIPTION
**Pull request by brizental**
The second commit (008bce4) is a bonus. I am not sure that is a good pattern, but it does centralize payload validation instead of doing it in a bunch of different places and in different ways, and I think that at least is good. Please let me know what you think :)

### CODE REVIEW

File: src/database.ts [hunk 1]
```diff
@@ -110,7 +110,7 @@ class Database {
    * @param metric The metric to record to.
    * @param transformFn The transformation function to apply to the currently persisted value.
    */
-  async transform(metric: Metric, transformFn: (v?: MetricPayload) => MetricPayload): Promise<void> {
+  async transform<T extends MetricPayload>(metric: Metric, transformFn: (v?: T) => T): Promise<void> {
```
Comment (Dexterp37) [inline src/database.ts:113]: Good call.

File: src/metrics/counter.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  if (v <= 0) {
+    return false;
+  }
+
+  return true;
+}
+
+class CounterMetric extends Metric {
```
Comment (Dexterp37) [inline src/metrics/counter.ts:39]: I now noticed that we're missing the class documentation. Can you kindly either file a bug or add it for this and the other metrics as well? Just copy over and adapt the one from the SDK

File: src/metrics/counter.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  constructor(meta: CommonMetricData) {
+    super("counter", meta);
```
Comment (Dexterp37) [inline src/metrics/counter.ts:41]: I suspect we can do better than storing a string for each metric. Can you please file a bug about this?

File: src/metrics/counter.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    if (amount <= 0) {
+      // TODO: record error once Bug 1682574 is resolved.
+      console.warn(`Attempted to add an invalid amount ${amount}. `);
+      return;
+    }
+
+    const transformFn = ((amount) => {
+      return (v?: CounterMetricPayload): CounterMetricPayload =>  {
+        let result = v ? v + amount : amount;
+        if (result > Number.MAX_SAFE_INTEGER) {
```
Comment (Dexterp37) [inline src/metrics/counter.ts:74]: Does this change the maximum representable number compared to the SDK?

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: src/metrics/counter.ts (hunk 2 above)
Comment (brizental) [reply src/metrics/counter.ts:39]: SHISH, you are right. Nah, I can do it in this PR, no need for yet another bug.

File: src/metrics/counter.ts (hunk 3 above)
Comment (brizental) [reply src/metrics/counter.ts:41]: We can have an enum (which will just compile down to a list of strings), like for Lifetime (see: https://github.com/mozilla/glean.js/blob/main/src/metrics/index.ts#L10-L17). Do you think that is better?

Comment (Dexterp37) [reply src/metrics/counter.ts:41]: Let's discuss about this in a separate bug. Storing one string per metric instance could be problematic, when it comes to memory consumption of the library. We should measure this along with the overall performances of Glean.js.

File: src/metrics/counter.ts (hunk 4 above)
Comment (brizental) [reply src/metrics/counter.ts:74]: Good question, let me check... Actually is much larger then the maximum representable number on the SDK:

- Javascript: 9007199254740991
- Rust: 4294967295

File: src/metrics/counter.ts (hunk 3 above)
Comment (brizental) [reply src/metrics/counter.ts:41]: FWIW now that I think about we can also just remove the type have `instaceof` checks for checking the metric type. I'll file an investigation bug.

Comment (brizental) [reply src/metrics/counter.ts:41]: https://bugzilla.mozilla.org/show_bug.cgi?id=1682864

**Review by Dexterp37 — APPROVED** [review]

**Pull request merged**